channel/state.go

```diff
@@ -104,20 +110,16 @@ func (s *State) Clone() *State {
 
 // Encode encodes a state into an `io.Writer` or returns an `error`.
 func (s State) Encode(w io.Writer) error {
-	err := perunio.Encode(w, s.ID, s.Version, s.Allocation, s.IsFinal, OptAppEnc{s.App}, s.Data)
-	return errors.WithMessage(err, "state encode")
+	return errors.WithMessage(
+		perunio.Encode(w, s.ID, s.Version, s.Allocation, s.IsFinal, OptAppAndDataEnc{s.App, s.Data}),
```

Comment:
Why not use same struct for encoding and decoding?

Reply:
It can be done. But, I have not changed it as the scope of this PR is only to convert Data, Action from Encoders to binary Data type. 

Since, this PR in itself is a bit large, my suggestion would be to do that change in another PR. What do you think ?